### PR TITLE
Added data loader

### PR DESCRIPTION
**Pull request by hungngo97**
#6 Fix to use data loader for external links

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "supertest": "^4.0.2",
     "ts-jest": "^25.4.0",
     "ts-node": "^8.9.1",
-    "typescript": "^3.8.3"
+    "typescript": "^3.8.3",
+    "vega-loader": "^4.2.2"
```
Comment (domoritz) [inline package.json:43]: vega-loader is already coming in through Vega

File: src/app.ts [hunk 2]
```diff
@@ -55,8 +60,21 @@ app.post('/', async (req: Request, res: Response) => {
         .end('Invalid Schema, should be Vega or Vega-Lite.');
   }
 
-  const view = new vega.View(vega.parse(specs), {
+  let loader = vega.loader({ mode: 'http'});
+  loader.http = (uri:string, options:any): Promise<string> => {
+    if (ALLOWED_URLS.every((allowedUrl) => !uri.includes(allowedUrl))) {
+      res.status(404).send("External URI not allowed on this API");
+      throw 'External data url not allowed';
```
Comment (domoritz) [inline src/app.ts:67]:
```suggestion
      throw new Error('External data url not allowed');
```

File: src/app.ts [hunk 3]
```diff
@@ -55,8 +60,21 @@ app.post('/', async (req: Request, res: Response) => {
...
+    }
+    return fetch(uri, {}).then((response) => {
+      if (!response.ok) throw response.status + '' + response.statusText;
```
Comment (domoritz) [inline src/app.ts:70]:
```suggestion
      if (!response.ok) throw new Error(response.status + response.statusText);
```

File: test/index.test.ts [hunk 4]
```diff
@@ -24,4 +25,9 @@ describe('API Request', () => {
     expect(response.statusCode).toBe(200);
     expect(response.text).toMatchSnapshot();
   });
+
+  test('should return 404 for external link', async () => {
```
Comment (domoritz) [inline test/index.test.ts:29]: Why 404? It should be something like "not allowed".

File: src/app.ts [hunk 5]
```diff
@@ -55,8 +60,21 @@ app.post('/', async (req: Request, res: Response) => {
         .end('Invalid Schema, should be Vega or Vega-Lite.');
   }
 
-  const view = new vega.View(vega.parse(specs), {
+  let loader = vega.loader({ mode: 'http'});
+  loader.http = (uri:string, options:any): Promise<string> => {
+    if (ALLOWED_URLS.every((allowedUrl) => !uri.includes(allowedUrl))) {
```
Comment (domoritz) [inline src/app.ts:65]: includes is not enough. `http://www.google.com#vega.github.io` if `vega.github.io` is allowed

Comment (hungngo97) [reply src/app.ts:65]: How about `startsWith('vega.github.io')` ?

Comment (domoritz) [reply src/app.ts:65]: Use https://developer.mozilla.org/en-US/docs/Web/API/URL instead

File: src/app.ts [hunk 6]
```diff
@@ -56,18 +57,19 @@ app.post('/', async (req: Request, res: Response) => {
...
-      res.status(404).send("External URI not allowed on this API");
-      throw 'External data url not allowed';
+    let parsedUri = new URL(uri);
+    if (ALLOWED_URLS.every((allowedUrl) => !parsedUri.hostname.includes(allowedUrl))) {
+      res.status(400).send("External URI not allowed on this API");
+      throw new Error('External data url not allowed');
     }
     return fetch(uri, {}).then((response) => {
-      if (!response.ok) throw response.status + '' + response.statusText;
+      if (!response.ok) throw new Error(response.status + '' + response.statusText);
```
Comment (domoritz) [inline src/app.ts:72]:
```suggestion
      if (!response.ok) throw new Error(response.status + response.statusText);
```

File: src/constants.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export const ALLOWED_URLS: string[] = ['www.vega.github.io'];
```
Comment (domoritz) [inline src/constants.ts:1]:
```suggestion
export const ALLOWED_URLS: string[] = ['vega.github.io'];
```

**Pull request merged**